### PR TITLE
#46-Update-Controller

### LINKED ISSUE #46: Off Screen - Update Controller
### Description

We need a controller for the signals and slots the frontend needs to display its information

### Acceptance Criteria

- [ ] Information on motor temp, pack temp, charge state
- [ ] Status’ for side brbs bms imd... (look at mock) Should probably use an array or dictionary to track statuses so that you can easily add or remove statuses. 

### Proposed Solution

- [ ] QProperties 

### Mocks

Screenshot 2023-12-16 at 4 09 32 PM

### PR DESCRIPTION
**Pull request by mliam0608**
Added a signalsController.cpp and signalsController.h file. The new signalsController class can check and change the status for different signals

## Changes

Added a SignalsController class inside the controllers folder.

## Notes

I have not tested the code yet.

Closes #46

### CODE REVIEW

File: NERODevelopment/src/controllers/signalsController.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#define SIGNALSCONTROLLER_H
+#include <QObject>
+#include <QMap>
+#include <QString>
+
+/**
+ * @brief The TheSignalsController class
+ * Controller for the signals view
+ */
+class SignalsController : public QObject {
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/signalsController.h]: Rename to OffViewController

File: NERODevelopment/src/controllers/signalscontroler.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#include "signalscontroler.h"
+
+signalsController::signalsController(QObject *parent)
+{
+    // Inital mockup of statuses, Strings represent
+    // the name and status of each attribute
+    attributeStatus["SideBRBs"] = "unchecked";
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/signalscontroler.cpp]: can you make these macros

File: NERODevelopment/src/controllers/signalsController.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    Q_OBJECT
+    Q_PROPERTY(QMap<QString, QString> attributeStatus READ getAttributeStatus NOTIFY attributeStatusChanged)
+public:
+    explicit SignalsController(QObject *parent = nullptr);
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/signalsController.h]: you also need properties for pack temp, motor temp and state of charge

File: NERODevelopment/src/controllers/signalscontroler.cpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    attributeStatus["BMS"] = "unchecked";
+    attributeStatus["IMD"] = "unchecked";
+    attributeStatus["BSPD"] = "unchecked";
+    attributeStatus["MPU"] = "unchecked";
+    attributeStatus["BOTS"] = "unchecked";
+    attributeStatus["INERTIA"] = "unchecked";
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/signalscontroler.cpp]: also the unchecked should be an enumeration of possible values, so like good, faulted, off etc

Comment (Peyton-McKee) [reply NERODevelopment/src/controllers/signalscontroler.cpp]: you can do that through macros in a file

File: NERODevelopment/src/controllers/signalscontroler.cpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    attributeStatus["HV CNCTR"] = "unchecked";
+
+}
+
+QMap<QString, QString> SignalsController::getAttributeStatus(const QString &attribute) const
+{
+    return attributeStatus.value(attribute);
+}
+
+void SignalsController::setAttributeStatus(const QString &attribute, const QString &status)
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/signalscontroler.cpp]: you need to make an update function that calls the models get functions for all of these values and sets the attribute status for all of them. Your probably going to need to make the model functions for that as well

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
Good start, Make sure to rename the file to OffViewController

File: NERODevelopment/src/controllers/signalsController.h (hunk 1 above)
Comment (mliam0608) [reply NERODevelopment/src/controllers/signalsController.h:11]: Done!

File: NERODevelopment/src/controllers/signalscontroler.cpp (hunk 2 above)
Comment (mliam0608) [reply NERODevelopment/src/controllers/signalscontroler.cpp:7]: I added these in AttributeStatus.h

File: NERODevelopment/src/controllers/signalsController.h (hunk 3 above)
Comment (mliam0608) [reply NERODevelopment/src/controllers/signalsController.h]: I added pointers to reference these properties from the other files. Does that work?

File: NERODevelopment/src/controllers/signalscontroler.cpp (hunk 4 above)
Comment (mliam0608) [reply NERODevelopment/src/controllers/signalscontroler.cpp:13]: Done!

File: NERODevelopment/src/controllers/offViewController.cpp [hunk 6]
```diff
@@ -0,0 +1,28 @@
+#include "offViewController.h"
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/offViewController.cpp]: this file doesnt follow current naming conventions

File: NERODevelopment/src/controllers/attributeNames.h [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#ifndef ATTRIBUTENAMES_H
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/attributeNames.h]: file doesnt follow naming conventions

File: NERODevelopment/src/controllers/attributeNames.h [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+#define ATTRIBUTENAMES_H
+
+#define SIDEBRBS "SideBRBs"
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/attributeNames.h]: All caps

File: NERODevelopment/src/controllers/attributestatus.h [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#ifndef ATTRIBUTESTATUS_H
+#define ATTRIBUTESTATUS_H
+
+enum class AttributeStatus {
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/attributestatus.h]: assign raw values to them, like = 0, 1, 2

File: NERODevelopment/src/controllers/offviewmodel.cpp [hunk 10]
```diff
@@ -0,0 +1,13 @@
+#include "OffViewModel.h"
+
+OffViewModel::OffViewModel(QObject *parent) : QObject(parent)
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/offviewmodel.cpp]: no this file shouldnt exist, you should add getters to the current model for all of these values and add them to the current macro file for the model

File: NERODevelopment/src/controllers/offviewmodel.h [hunk 11]
```diff
@@ -0,0 +1,18 @@
+#ifndef OFFVIEWMODEL_H
+#define OFFVIEWMODEL_H
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/offviewmodel.h]: same with this

File: NERODevelopment/src/controllers/signalsController.h (hunk 1 above)
Comment (Peyton-McKee) [reply NERODevelopment/src/controllers/signalsController.h]: Still this, you are creating one controller

File: NERODevelopment/src/controllers/offviewmodel.cpp [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+{
+    for (const auto& attributeName : {SIDEBRBS, BMS, BSPD, MPU, BOTS, INERTIA, CPBRB, TSMS, HVDINTRLK, HVCNCTR}) {
+        m_attributeStatusMap[attributeName] = AttributeStatus::OFF;
+    }
+}
+
+AttributeStatus OffViewModel::getAttributeStatus(AttributeName attributeName) const
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/offviewmodel.cpp]: if anything make this a function in the actual model

File: NERODevelopment/src/controllers/signalsController.h (hunk 3 above)
Comment (Peyton-McKee) [reply NERODevelopment/src/controllers/signalsController.h]: where

File: NERODevelopment/src/controllers/signalscontroler.cpp (hunk 2 above)
Comment (Peyton-McKee) [reply NERODevelopment/src/controllers/signalscontroler.cpp]: you didnt actually use the macros?

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
You should be only making one controller. You need to delete the signals controller if your not using it anymore, also you still need to add pack, motor, state of charge as other q properties. Make sure to tag your commits with the issue number, and also dont make a new model, just ad the function to the current models

File: NERODevelopment/src/controllers/offViewController.cpp [hunk 13]
```diff
@@ -0,0 +1,54 @@
+#include "offviewcontroller.h"
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/offViewController.cpp]: follow file naming conventions

File: NERODevelopment/src/controllers/offViewController.h [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+                   motorTempChanged FINAL)
+    Q_PROPERTY(int stateOfCharge READ stateOfCharge WRITE setStateOfCharge NOTIFY
+                   stateOfChargeChanged FINAL)                                              //QProperty for state of charge
+
+public:
+    explicit OffViewController(OffViewModel *viewModel, QObject *parent = nullptr);
+
+    QMap<QString, AttributeStatus> getAttributeStatus() const;
+    int packTemp() const;
+    qreal motorTemp() const;
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/offViewController.h]: we use floats i think

File: NERODevelopment/src/controllers/offViewController.cpp [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    if (m_attributeStatus.value(name) != status)
+    {
+        m_attributeStatus[name] = status;
+        emit attributeStatusChanged(m_attributeStatus);
+    }
+}
+
+void OffViewController::updateAttributesFromModel() {
+    for (const auto& attributeName : {SIDEBRBS, BMS, BSPD, MPU, BOTS, INERTIA, CPBRB, TSMS, HVDINTRLK, HVCNCTR}) {
+        setAttributeStatus(attributeName, m_attributeStatus[attributeName]);
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/offViewController.cpp]: this is just going to set it to itself isnt it? You need to get the updated value from the model. You can use the getById Function I believe. Actually you can add the attribute names macros to the data_type_names.h macro file

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
you need to connect the model and the controller, and delete the viewmodel attribute, also move the macros to the data_type_name file

File: NERODevelopment/src/controllers/offviewcontroller.cpp [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    if (m_attributeStatus.value(name) != status)
+    {
+        m_attributeStatus[name] = status;
+        emit attributeStatusChanged(m_attributeStatus);
+    }
+}
+
+void OffViewController::updateAttributesFromModel() {
+    const auto attributeStatusMap = m_model->getAttributeStatus();
+    for (const auto& attributeName : {SIDEBRBS, BMS, BSPD, MPU, BOTS, INERTIA, CPBRB, TSMS, HVDINTRLK, HVCNCTR}) {
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/offviewcontroller.cpp:22]: you also need to update the pack temp, motor temp and state of charge, the model already has dedicated functions for these

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
one last thing then you should be good

File: NERODevelopment/src/controllers/offviewcontroller.cpp [hunk 17]
```diff
@@ -0,0 +1,53 @@
+#include "offviewcontroller.h"
+
+OffViewController::OffViewController(Model *model, QObject *parent)
+    : QObject{parent}, m_packTemp(0.0), m_motorTemp(0.0), m_stateOfCharge(0.0) {
+    this->m_model = model;
+    connect(m_model, &Model::updateAttributesFromModel, this,
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/offviewcontroller.cpp]: this compiles? i dont think this function exists in the model, you want the models currentDataDidChange Signal

File: NERODevelopment/src/controllers/attributestatus.h [hunk 18]
```diff
@@ -0,0 +1,10 @@
+#ifndef ATTRIBUTESTATUS_H
```
Comment (Peyton-McKee) [inline NERODevelopment/src/controllers/attributestatus.h]: move this file to utils

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
You have to add these files to the actual project in the cmakelists.txt file under project sources. Then when you try to build youll see that it fails

**Review by Peyton-McKee — APPROVED** [review]
Beautiful

**Pull request merged**